orderer/common/channelparticipation/restapi.go

```diff
@@ -127,17 +132,67 @@ func (h *HTTPHandler) serveListOne(resp http.ResponseWriter, req *http.Request)
 	h.sendResponseOK(resp, infoFull)
 }
 
-// Join a channel
+// Join a channel. Expect application/json content.
 func (h *HTTPHandler) serveJoin(resp http.ResponseWriter, req *http.Request) {
 	_, err := negotiateContentType(req) // Only application/json for now
 	if err != nil {
 		h.sendResponseJsonError(resp, http.StatusNotAcceptable, err)
 		return
 	}
 
-	//TODO
-	err = errors.Errorf("not implemented yet: %s %s", req.Method, req.URL.String())
-	h.sendResponseJsonError(resp, http.StatusNotImplemented, err)
+	channelID, ok := mux.Vars(req)[channelIDKey]
+	if !ok {
+		h.sendResponseJsonError(resp, http.StatusInternalServerError, errors.Wrap(err, "missing channel ID"))
+		return
+	}
+
+	if err = configtx.ValidateChannelID(channelID); err != nil {
+		h.sendResponseJsonError(resp, http.StatusBadRequest, errors.Wrap(err, "invalid channel ID"))
+		return
+	}
+
+	bodyBytes, err := ioutil.ReadAll(req.Body)
+	if err != nil {
+		h.logger.Debugf("Failed to read request body: %s", err)
+		h.sendResponseJsonError(resp, http.StatusBadRequest, errors.Wrap(err, "cannot read request body"))
+		return
+	}
+
+	body := types.JoinBody{}
+	err = json.Unmarshal(bodyBytes, &body)
```

Comment:
That assumes you augment config block into json, while multipart content type allows you just to insert the binary without it. For instance in the way you suggesting to do it, it will be very hard to utilize curl or wget for instance to submit join channel request.

Reply:
You are correct. Adding Content-type application/octet-stream and multipart/form-data will be added later. The plan is to support all 3 types, see:
https://jira.hyperledger.org/browse/FAB-17891